### PR TITLE
[Feat] Cleanup ModelCheckpoint / EarlyStopping by moving logic to LoggerConnector

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces MetricsHolder class which is responsible to hold metrics and their convert functions.

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
this is based on the 1.2 branch but unfortunately the Metrics code that needs cleanup is on master, and so this model checkpoint code will now diverge

**Comment by tchaton** [comment]
> 1.2 branch but unfortunately the Metrics code that needs cleanup is on master, and so this model checkpoint code will now diverge

Yes, I agree. But we won't introduce this kind of refactoring during Chrismas vacation. I will move this back to master in JAN, when the entire team is back.

Best,
T.C

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -25,25 +25,71 @@
...
 
 
 class LoggerConnector:
     def __init__(self, trainer):
         self.trainer = trainer
-        self.callback_metrics = {}
-        self.evaluation_callback_metrics = {}
-        self.logged_metrics = {}
-        self.progress_bar_metrics = {}
+        self._callback_metrics = MetricsHolder()
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: Are all these only for internal use?

File: pytorch_lightning/core/step_result.py [hunk 2]
```diff
@@ -367,7 +367,10 @@ def get_forked_metrics(self, add_dataloader_idx=False):
             dl_key = self._add_dataloader_idx(k, options["dataloader_idx"], add_dataloader_idx)
 
             if options['forked']:
-                result[dl_key] = self[k]
+                if isinstance(self[k], Metric):
+                    result[dl_key] = self[k].compute().detach()
```
Comment (SkafteNicki) [inline pytorch_lightning/core/step_result.py:371]: I cannot find the `detach()` in the new class? Is it not needed anymore?

**Review by SkafteNicki — COMMENTED** [review]
PR looking good. 
@tchaton just to clarify, this is only refactor and no new features right?

File: pytorch_lightning/core/step_result.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/core/step_result.py:371]: Good catch ! I forgot it.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: Yes, I moved them for internal use only now. The exposed ones are the converted internals.

**Comment by tchaton** [comment]
> just to clarify, this is only refactor and no new features right?

Hey @SkafteNicki, it is mainly a refactor + a small feature (automatic casting when user access a metric dictionary).

Reason: Users had to add conversion check within callbacks which they shouldn't have to.

This PR adds automatic conversion only on accessing properties from the trainer. For performance reasons, I moved all the metrics to private, so they don't trigger the conversion.

Best,
T.C

**Comment by tchaton** [comment]
Hey @SkafteNicki, It enables having all the metrics displayed in the same format.

Screenshot 2020-12-28 at 11 50 06

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import numbers
+from typing import Any
+
+import torch
+
+from pytorch_lightning.metrics.metric import Metric
+from pytorch_lightning.utilities import _TPU_AVAILABLE
+
+
+class MetricsHolder:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:23]: You probably need unit tests for this class and its methods :)

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return self.metrics.pop(key, default)
+
+    def reset(self, metrics):
+        self.metrics = metrics
+
+    def convert(self, use_tpu: bool, device: torch.device):
+        for key, value in self.metrics.items():
+            self.metrics[key] = self._convert(value, use_tpu, device)
+
+    def _convert_to_float(self, current, use_tpu: bool, device: torch.device):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:51]: can we not just simplify as convert_to_float(x) = float(convert_to_tensor(x)) ?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -31,19 +32,64 @@
...
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: I very much agree with this protection, but in such case we probably need to keep API back-compatible:
- add the past as property with a warning whenever used
- include in a warning deprecation warning
- add tests that the separation warning is properly called

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -31,19 +32,64 @@
...
+        self._logged_metrics = MetricsHolder()
+        self._progress_bar_metrics = MetricsHolder()
         self.eval_loop_results = []
         self._cached_results = {stage: EpochResultStore(trainer, stage) for stage in LoggerStages}
         self._callback_hook_validator = CallbackHookNameValidator()
         self._current_stage = None
 
+    @property
+    def callback_metrics(self):
+        return self.get_metrics("callback_metrics")
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:46]: add depreactions?

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    It holds metris and implement convertion functions.
+    Those functions will be triggered within LoggerConnector
+    when the property is being requested from the user.
+    """
+
+    def __init__(self, to_float: bool = False):
+        self.metrics = {}
+        self._convert = self._convert_to_tensor
+        if to_float:
+            self._convert = self._convert_to_float
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:36]: can we avoid dynamically bound methods? 
It would be more readable and debuggable if we just use the appropriate function in the convert method down below.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 8]
```diff
@@ -31,19 +32,64 @@
...
+    @callback_metrics.setter
+    def callback_metrics(self, callback_metrics):
+        self.set_metrics("callback_metrics", callback_metrics)
+
+    @property
+    def evaluation_callback_metrics(self):
+        return self.get_metrics("evaluation_callback_metrics")
+
+    @evaluation_callback_metrics.setter
+    def evaluation_callback_metrics(self, evaluation_callback_metrics):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:57]: can we add typing?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: It is still backward compatible as I moved them as property.

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:36]: Sounds great ! Let me do so 👍

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:23]: Yes, I will do so :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: yes, but do we deprecate it in the future?

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:51]: Done

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:36]: done

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:46]: Not sure to understand the need for `depreactions`. Nothing changed from the user facing point of view.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:57]: Yes.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: I don't understand why you think we should deprecate something.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:35]: It basically just add property to LoggerConnector, it calls convert every time the user request the metrics and don't when we gather metrics internally as moved into private variable.

File: tests/trainer/logging/test_logger_connector.py [hunk 9]
```diff
@@ -15,22 +15,25 @@
...
 from pytorch_lightning.trainer.connectors.logger_connector.callback_hook_validator import CallbackHookNameValidator
+from pytorch_lightning.trainer.connectors.logger_connector.metrics_holder import MetricsHolder
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from tests.base.boring_model import BoringModel, RandomDataset
 
 
-def decorator_with_arguments(fx_name='', hook_fx_name=None):
-    def decorator(func):
-        def wrapper(self, *args, **kwargs):
+def decorator_with_arguments(fx_name: str = '', hook_fx_name: str = None) -> callable:
```
Comment (awaelchli) [inline tests/trainer/logging/test_logger_connector.py:34]: I wonder why it is not failing on import? should it be Callable from typing?

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/logging/test_logger_connector.py (hunk 9 above)
Comment (tchaton) [reply tests/trainer/logging/test_logger_connector.py]: You are right ! Change to Callable from typing.

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
+# Copyright The PyTorch Lightning team.
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:1]: shall this rather go to metric sub-package?

Comment (justusschock) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:1]: I think this is fine here since this is specifically for logging and metrics shall not be entangled to the rest of lightning

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 11]
```diff
@@ -605,7 +602,7 @@ def _update_best_and_save(
             del_list.append(delpath)
 
         # do not save nan, replace with +/- inf
-        if torch.isnan(current):
+        if isinstance(current, torch.Tensor) and torch.isnan(current):
             current = torch.tensor(float('inf' if self.mode == "min" else '-inf'))
```
Comment (justusschock) [inline pytorch_lightning/callbacks/model_checkpoint.py:606]: should this always be on cpu? or should it be on `current.device`?

**Review by justusschock — APPROVED** [review]
LGTM just one minor question regarding device.

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:1]: It isn't related to Metric. It is just a Dictionary Holder for LoggerConnector.

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 12]
```diff
@@ -375,7 +375,7 @@ def update_logger_connector(self) -> None:
 
             if is_train:
                 # Only log and add to callback epoch step during evaluation, test.
-                logger_connector.logged_metrics.update(batch_log_metrics)
+                logger_connector._logged_metrics.update(batch_log_metrics)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:378]: Great catch ! Thanks !

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:606]: Great question ! I am not sure. What do you think ?

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:1]: Could change the class name.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+from pytorch_lightning.metrics.metric import Metric
+from pytorch_lightning.utilities import _TPU_AVAILABLE
+
+
+class MetricsHolder:
+
+    """
+    This class acts as a dictonary holder.
+    It holds metris and implement convertion functions.
```
Comment (s-rog) [inline pytorch_lightning/trainer/connectors/logger_connector/metrics_holder.py:27]:
```suggestion
    It holds metrics and implements conversion functions.
```

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 11 above)
Comment (justusschock) [reply pytorch_lightning/callbacks/model_checkpoint.py:585]: IMO it should live on current.device, since all the other tensors (especially current if not nan) also live on this device.
